runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,573 @@
+  }
+  state.buffers[bufnr] = buf_state
+
+  api.nvim_buf_attach(bufnr, false, {
+    on_lines = function()
+      -- detach if the LSP client detached from this buffer or on_detach() was
+      -- called manually
+      local s, bs = get_state(bufnr, client_id)
+      if not s or not bs or not vim.lsp.buf_is_attached(bufnr, client_id) then
+        return true
+      end
+    end,
+    on_reload = function()
+      api.nvim_buf_clear_namespace(bufnr, state.namespace, 0, -1)
+      reset_buf_state(bufnr, client, true)
+      request_all_tokens(bufnr, client_id)
+    end,
+    on_detach = function()
+      M.stop(bufnr, client_id)
+    end,
+  })
+
+  -- Add a BufWinEnter event to re-highlight buffers whose state was reset
+  -- without a document version change, such as after receiving a refresh
+  -- request from the server while the buffer was hidden
+  local augroup_str = ('vim_lsp_c_%d_b_%d_semantic_tokens'):format(client_id, bufnr)
+  local augroup = api.nvim_create_augroup(augroup_str, { clear = true })
+  api.nvim_create_autocmd({ 'BufWinEnter' }, {
+    buffer = bufnr,
+    group = augroup,
+    callback = function()
+      request_all_tokens(bufnr, client_id)
+    end,
+  })
+
+  request_all_tokens(bufnr, client_id)
+end
+
+--- Stop the semantic token highlighting engine for the given buffer with the
+--- given client.
+---
+--- NOTE: This is currently called automatically by |buf_detach_client|. You
+--- shouldn't need to call this function manually if semantic highlighting was
+--- started automatically by |buf_attach_client|.
+---
+---@param bufnr number
+---@param client_id number
+function M.stop(bufnr, client_id)
+  vim.validate({
+    bufnr = { bufnr, 'n', false },
+    client_id = { client_id, 'n', false }
+  })
+
+  local client = vim.lsp.get_client_by_id(client_id)
+  if not client then
+    return
+  end
+
+  local state, buf_state = get_state(bufnr, client_id)
+  if not state or not buf_state then
+    return
+  end
+
+  api.nvim_buf_clear_namespace(bufnr, state.namespace, 0, -1)
+  reset_buf_state(bufnr, client, false)
+
+  state.buffers[bufnr] = nil
+  if #state.buffers == 0 then
+    state_by_client[client_id] = nil
+  end
+
+  for _, test in pairs(vim.lsp.buf_get_clients(bufnr)) do
+    if client_id ~= test.id then
+      return
+    end
+  end
+  active_buffers[bufnr] = nil
+end
+
+--- Used by the `textDocument/didChange` handler to notify the semantic token
+--- highlighting engine that a change was sent to the LSP server. Currently it
+--- will just send new request for semantic tokens to the server.
+---
+---@private
+---@param bufnr number
+---@param client_id number
+function M.on_did_change(bufnr, client_id)
+  -- TODO: we want to keep as much of this API private as possible, but the main
+  -- lsp module needs access in its send_changes() function to be able to request
+  -- tokens on a buffer sync event. The alternative is to set up our own debounce
+  -- timer attached to on_lines, or wait for #15275.
```

Comment:
This is the other part I'm not 100% sure on. Two things about this I'm not a huge fan of:

1. Requires a callback when the `textDocument/didChange` notification is sent. #15275 would "fix" this and no longer require a special call to the semantic token module inside the `send_changes()` call in the main lsp module. This also means we have to have an API function accessible outside this module that sends the semantic tokens request.
2. Sending a semantic tokens request for every didChange notification can be excessive.

I allude to it in the TODO here, but we can do this a couple different ways.

1. We can manage our own debounce timer that we can start in `on_lines`. I have an alternate branch that does this and it works well. We'd need a configuration option (per client? take it from cursor hold? not sure) for the actual debounce time. If we did this, I might also add a new autocmd for `InsertLeave` that cancels the timer and sends the request immediately.
2. We can add more autocmd hooks for things like `CursorHold`, `CursorHoldI`, etc and not worry about timers at all. This is a bit less flexible and could cause more delay than actually necessary in getting new tokens.

Reply:
I opted for a configurable debounce timer that defaults to 200ms. It's configurable via an optional kwargs field in `start()`.  Feels cleaner to not need a specific hook in the didChange notification code.

Added `InsertLeave` to autocmds that send a new request so the debounce is cancelled and the request is sent immediately via n that event.